tools/checky/src/main.rs

```diff
@@ -0,0 +1,51 @@
+use std::path::PathBuf;
+
+use checks::check;
+use clap::Parser;
+use color_eyre::{eyre::Context, Result};
+use rust_file::RustFile;
+use walkdir::WalkDir;
+
+mod checks;
+mod rust_file;
+mod syn_context;
+
+#[derive(Parser, Debug)]
+struct Arguments {
+    paths: Vec<PathBuf>,
+}
+
+fn main() -> Result<()> {
+    let arguments = Arguments::parse();
+
+    for path in arguments.paths {
+        for entry in WalkDir::new(&path) {
+            let entry =
+                entry.wrap_err_with(|| format!("failed to walk directory {}", path.display()))?;
+
+            if !entry.file_type().is_file() {
+                continue;
+            }
+
+            if !entry
+                .path()
+                .extension()
+                .map(|extension| extension == "rs")
+                .unwrap_or_default()
+            {
+                continue;
+            }
+
+            let file = RustFile::try_parse(entry.path())
+                .wrap_err_with(|| format!("failed to parse {}", entry.path().display()))?;
+
+            for report in check(&file) {
```

Comment:
```suggestion
            let reports = check(&file);
            for report in reports {
```

Just because this is where the magic happens...

Reply:
Done